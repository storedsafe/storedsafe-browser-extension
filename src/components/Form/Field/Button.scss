@import '../../../variables';

.field-button .field-control {
  border: 2px solid $primary;
  border-radius: $border-radius;
  background-color: $primary;
  color: $bg;
  font-weight: 800;
  text-transform: uppercase;

  &:hover,
  &:focus,
  &:active {
    color: $accent;
    box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.1);
  }

  &:hover {
    border-color: $border-hover;
    background-color: $bg-hover;
  }

  &:focus {
    border-color: $border-focus;
    background-color: $bg-focus;
  }

  &:active {
    border-color: $border-active;
    background-color: $bg-active;
  }
}
