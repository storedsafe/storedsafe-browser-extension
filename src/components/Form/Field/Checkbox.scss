@import '../../../variables';

.field-checkbox {
  $size: 20px;
  $border-size: 2px;
  $padding: 2px;
  $inner-size: $size - ($border-size * 2) - ($padding * 2);

  .field-container {
    display: flex;
    flex-direction: column;
    border: 0;
  }

  label {
    display: flex;
    align-items: center;
  }

  input {
    height: 0;
    opacity: 0;
    position: absolute;
    width: 0;
  }

  .custom-checkbox {
    padding: $padding;
    position: relative;
    border: $border-size solid $border;
    border-radius: $border-radius;
    box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1) inset;
    width: $size;
    height: $size;
    margin-right: 0.5em;
  }

  .custom-checkbox::before {
    display: none;
    position: absolute;
    content: '';
    width: $inner-size;
    height: $inner-size;
    background-color: $accent;
    box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1);
    border-radius: $border-radius;
  }

  input:checked ~ .custom-checkbox::before {
    display: block;
  }

  input:hover ~ .custom-checkbox,
  input:active ~ .custom-checkbox,
  input:focus ~ .custom-checkbox {
    box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1) inset, 0 0 3px 0 rgba(0, 0, 0, 0.1);
  }

  input:hover ~ .custom-checkbox {
    border-color: $border-hover;
    background-color: $bg-hover;
  }

  input:focus ~ .custom-checkbox {
    border-color: $border-focus;
    background-color: $bg-focus;
  }

  input:active ~ .custom-checkbox {
    border-color: $border-active;
    background-color: $bg-active;
  }
}
