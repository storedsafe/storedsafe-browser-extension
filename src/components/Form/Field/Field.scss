@import '../../../variables';

.field-label {
  text-transform: uppercase;
  user-select: none;
}

.field-control {
  border: 0;
  border-bottom: 2px solid $border;
  padding: $input-spacing;
  box-sizing: border-box;
  outline: 0;
  width: 100%;
  transition: $transition-duration;

  &:hover,
  &:focus,
  &:active {
    box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.1);
  }

  &:hover {
    border-color: $border-hover;
    background-color: $bg-hover;
  }

  &:focus {
    border-color: $border-focus;
    background-color: $bg-focus;
  }
}

.field {
  label,
  input,
  textarea,
  select,
  option,
  button {
    font-size: 1rem;
    font-family: 'Roboto', sans-serif;
  }
}
