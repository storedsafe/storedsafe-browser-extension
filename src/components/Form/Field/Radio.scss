@import '../../../variables';

.field-radio {
  $size: 20px;
  $border-size: 2px;
  $padding: 2px;
  $inner-size: $size - ($border-size * 2) - ($padding * 2);

  .field-container {
    display: flex;
    flex-direction: column;
  }

  label {
    display: flex;
    align-items: center;
  }

  input {
    height: 0;
    opacity: 0;
    position: absolute;
    width: 0;
  }

  .custom-radio {
    padding: $padding;
    position: relative;
    border-radius: 100%;
    border: $border-size solid $border;
    box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1) inset;
    width: $size;
    height: $size;
    margin-right: 0.5em;
  }

  .custom-radio::after {
    display: none;
    position: absolute;
    border-radius: 100%;
    content: '';
    width: $inner-size;
    height: $inner-size;
    background-color: $accent;
  }

  input:checked ~ .custom-radio::after {
    display: block;
  }

  input:hover ~ .custom-radio,
  input:active ~ .custom-radio,
  input:focus ~ .custom-radio {
    box-shadow: 0 0 3px 0 rgba(0, 0, 0, 0.1) inset, 0 0 3px 0 rgba(0, 0, 0, 0.1);
  }

  input:hover ~ .custom-radio {
    border-color: $border-hover;
    background-color: $bg-hover;
  }

  input:focus ~ .custom-radio {
    border-color: $border-focus;
    background-color: $bg-focus;
  }

  input:active ~ .custom-radio {
    border-color: $border-active;
    background-color: $bg-active;
  }
}
