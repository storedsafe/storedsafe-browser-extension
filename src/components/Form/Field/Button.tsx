import * as React from 'react';
import { FieldProps, FieldAttributes } from './Field';
import './Button.scss';

type ButtonProps = Pick<FieldProps, 'name' | 'type' | 'value' | 'onBlur' | 'label'>;
export const Button: React.SFC<ButtonProps | FieldAttributes> = ({
  name,
  type,
  label,
  value,
  onBlur,
  ...attributes
}: ButtonProps) => (
  <div className="field field-button">
    <button
      className="field-container field-control"
      id={name}
      name={name}
      defaultValue={value as string}
      type={type as 'button' | 'submit' | 'reset'}
      onBlur={(): void => {
        onBlur && onBlur(value, name);
      }}
      {...attributes}>
      {label}
    </button>
  </div>
);
