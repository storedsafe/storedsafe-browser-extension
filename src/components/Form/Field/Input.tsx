import * as React from 'react';
import { FieldProps, FieldAttributes } from './Field';
import './Input.scss';

type InputProps = Pick<FieldProps, 'name' | 'type' | 'value' | 'onChange' | 'onBlur' | 'label'>;
export const Input: React.SFC<InputProps | FieldAttributes> = ({
  name,
  type,
  value,
  label,
  onChange,
  onBlur,
  ...attributes
}: InputProps) => (
  <div className={`field field-input field-input-${type}`}>
    <label htmlFor={name}>
      <span className="field-label">{label}</span>
      <div className="field-container">
        <input
          className="field-control"
          id={name}
          name={name}
          type={type}
          value={['number', 'range'].includes(type) && isNaN(value as number) ? '' : value as string | number}
          onChange={({ target }: React.ChangeEvent<HTMLInputElement>): void => {
            if (['number', 'range'].includes(type)) {
              onChange && onChange(target.valueAsNumber, name);
            } else {
              onChange && onChange(target.value, name);
            }
          }}
          onBlur={({ target }: React.FocusEvent<HTMLInputElement>): void => {
            if (['number', 'range'].includes(type)) {
              onBlur && onBlur(target.valueAsNumber, name);
            } else {
              onBlur && onBlur(target.value, name);
            }
          }}
          {...attributes}
        />
      </div>
    </label>
  </div>
);

Input.defaultProps = {
  type: 'text',
  value: '',
}
