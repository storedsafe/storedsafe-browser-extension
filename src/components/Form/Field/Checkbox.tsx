import * as React from 'react';
import { FieldProps, FieldAttributes } from './Field';
import './Checkbox.scss';

type CheckboxProps = Pick<FieldProps, 'name' | 'value' | 'onChange' | 'onBlur' | 'label' | 'labelDir'>
export const Checkbox: React.SFC<CheckboxProps | FieldAttributes> = ({
  name,
  value,
  label,
  labelDir,
  onChange,
  onBlur,
  ...attributes
}: CheckboxProps) => (
  <div className="field field-checkbox">
    <div className="field-container">
      <label htmlFor={name} className="field-label">
        {labelDir === 'left' && (<span className="field-label">{label}</span>)}
        <input
          id={name}
          name={name}
          type="checkbox"
          checked={value as boolean}
          onChange={({ target }: React.ChangeEvent<HTMLInputElement>): void => {
            onChange && onChange(target.checked, target.name);
          }}
          onBlur={({ target }: React.FocusEvent<HTMLInputElement>): void => {
            onBlur && onBlur(target.checked, target.name);
          }}
          {...attributes}
        />
        <div className="custom-checkbox" />
        {labelDir === 'right' && (<span className="field-label">{label}</span>)}
      </label>
    </div>
  </div>
);

Checkbox.defaultProps = {
  value: false,
  labelDir: 'right',
};
