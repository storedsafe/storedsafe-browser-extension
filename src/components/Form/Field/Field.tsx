import * as React from 'react';
import { OnChangeCallback, OnBlurCallback, InputType } from '../Form';
import { Input } from './Input';
import { Checkbox } from './Checkbox';
import { Select } from './Select';
import { TextArea } from './TextArea';
import { Radio } from './Radio';
import { Button } from './Button';
import './Field.scss';

export type FieldAttributes =
  Omit<React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, 'name' | 'name' | 'type' | 'value' | 'checked' | 'onChange' | 'onBlur'>
| Omit<React.DetailedHTMLProps<React.SelectHTMLAttributes<HTMLSelectElement>, HTMLSelectElement>, 'name' | 'name' | 'value' | 'onChange' | 'onBlur'>
| Omit<React.DetailedHTMLProps<React.TextareaHTMLAttributes<HTMLTextAreaElement>, HTMLTextAreaElement>, 'name' | 'name' | 'value' | 'onChange' | 'onBlur'>;

export interface Option {
  title: string;
  value: string | number;
}

export interface FieldProps {
  name: string;
  label: string;
  type?: string;
  labelDir?: 'left' | 'right';
  options?: Option[];
  value?: InputType;
  onChange?: OnChangeCallback;
  onBlur?: OnBlurCallback;
}

export const Field: React.SFC<FieldProps | FieldAttributes> = ({
  type,
  ...attributes
}: FieldProps) => {
  if (type === 'select') {
    return <Select
      {...attributes}
    />;
  }
  if (type === 'textarea') {
    return <TextArea
      {...attributes}
    />
  }
  if (type === 'checkbox') {
    return <Checkbox
      {...attributes}
    />
  }
  if (type === 'radio') {
    return <Radio
      {...attributes}
    />
  }
  if (['button', 'submit', 'reset'].includes(type)) {
    return <Button
      type={type}
      {...attributes}
    />
  }

  return <Input
    type={type}
    {...attributes}
  />;
};

Field.defaultProps = {
  type: 'text',
}
