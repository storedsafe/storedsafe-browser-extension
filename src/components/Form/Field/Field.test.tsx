import * as React from 'react';
import { Field, Option } from './Field';
import { mount } from 'enzyme';

const options: Option[] = [
  { title: 'Foo', value: 'foo' },
  { title: 'Bar', value: 'bar' },
];

test('<Input />', () => {
  const wrapper = mount(<Field name="test" label="Text" />);
  const input = wrapper.find('input#test');
  expect(input.length).toBe(1);
  expect(input.props().type).toEqual('text');
  expect(wrapper).toMatchSnapshot();
});

test('<Input type=password />', () => {
  const wrapper = mount(<Field name="test" label="Text" type="password" />);
  const input = wrapper.find('input#test');
  expect(input.length).toBe(1);
  expect(input.props().type).toEqual('password');
  expect(wrapper).toMatchSnapshot();
});

test('<Checkbox />', () => {
  const wrapper = mount(<Field name="test" label="Text" type="checkbox" />);
  const input = wrapper.find('input#test');
  expect(input.length).toBe(1);
  expect(input.props().type).toEqual('checkbox');
  expect(wrapper).toMatchSnapshot();
});

test('<Radio />', () => {
  const wrapper = mount(<Field name="test" label="Text" type="radio" options={options} />);
  const input = wrapper.find('input[type="radio"]');
  expect(input.length).toBe(2);
  expect(input.length).toBe(2);
  expect(input.at(0).props().value).toEqual(options[0].value);
  expect(input.at(1).props().value).toEqual(options[1].value);
  expect(wrapper).toMatchSnapshot();
});

test('<TextArea />', () => {
  const wrapper = mount(<Field name="test" label="Text" type="textarea" />);
  const textarea = wrapper.find('textarea#test');
  expect(textarea.length).toBe(1);
  expect(wrapper).toMatchSnapshot();
});

test('<Select />', () => {
  const wrapper = mount(<Field name="test" label="Text" type="select" options={options} />);
  const select = wrapper.find('select#test');
  expect(select.length).toBe(1);
  const option = select.find('option');
  expect(option.length).toBe(2);
  expect(option.at(0).props().value).toEqual(options[0].value);
  expect(option.at(1).props().value).toEqual(options[1].value);
  expect(wrapper).toMatchSnapshot();
});

test('<Button />', () => {
  const wrapper = mount(<Field name="test" label="Text" type="button" />);
  const button = wrapper.find('button#test');
  expect(button.length).toBe(1);
  expect(wrapper).toMatchSnapshot();
});
