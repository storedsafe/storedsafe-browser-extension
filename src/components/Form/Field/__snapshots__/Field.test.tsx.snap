// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> 1`] = `ReactWrapper {}`;

exports[`<Checkbox /> 1`] = `ReactWrapper {}`;

exports[`<Input /> 1`] = `ReactWrapper {}`;

exports[`<Input type=password /> 1`] = `ReactWrapper {}`;

exports[`<Radio /> 1`] = `ReactWrapper {}`;

exports[`<Select /> 1`] = `ReactWrapper {}`;

exports[`<TextArea /> 1`] = `ReactWrapper {}`;
